universal/src/compression/zstd.cpp

```diff
@@ -0,0 +1,95 @@
+#include <userver/compression/zstd.hpp>
+
+#include <zstd.h>
+#include <zstd_errors.h>
+
+USERVER_NAMESPACE_BEGIN
+
+namespace compression::zstd {
+
+namespace {
+// The same size as in ZSTD_DStreamOutSize();
+const size_t kDecompressBufferSize = ZSTD_DStreamOutSize();
+}  // namespace
+
+std::string DecompressStream(std::string_view compressed, size_t max_size) {
+  std::string decompressed;
+  std::string buf(kDecompressBufferSize, '\0');
+
+  auto* stream = ZSTD_createDStream();
+  if (stream == nullptr) {
+    throw std::runtime_error("Couldn't create ZSTD decompression stream");
+  }
+
+  {
+    auto err_code = ZSTD_DCtx_reset(
+        stream, ZSTD_ResetDirective::ZSTD_reset_session_and_parameters);
+    if (ZSTD_isError(err_code)) {
+      ZSTD_freeDStream(stream);
+      throw ErrWithCode(ZSTD_getErrorName(err_code));
+    }
+  }
+
+  auto stream_del = [](ZSTD_DStream* ptr) { ZSTD_freeDStream(ptr); };
+  auto stream_guard =
```

Comment:
seems like you can move up this guard to line after ZSTD_createDStream
also ZSTD_freeDStream accept NULL pointer